ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Scheduling subsystem (in-OSTD part).
+//!
+//! This module defines what OSTD expects from a scheduling implementation
+//! and provides useful functions for controlling the execution flow.
+
+mod fifo_scheduler;
+
+use core::sync::atomic::{AtomicBool, Ordering};
+
+pub use fifo_scheduler::FifoScheduler;
+use spin::Once;
+
+use super::{processor, task::Task};
+use crate::{
+    arch::timer,
+    cpu::{local::PREEMPT_INFO, this_cpu},
+    prelude::*,
+};
+
+static SCHEDULER: Once<&'static dyn Scheduler<Task>> = Once::new();
+
+static SHOULD_PREEMPT: ShouldPreemptFlag = ShouldPreemptFlag::new();
+
+/// Injects a scheduler implementation into framework.
+///
+/// This function can only be called once and must be called during the initialization of kernel.
+pub fn inject_scheduler(scheduler: &'static dyn Scheduler<Task>) {
+    SCHEDULER.call_once(|| scheduler);
+
+    timer::register_callback(|| {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            if local_rq.update_current(UpdateFlags::Tick) {
+                SHOULD_PREEMPT.set();
+            }
+        })
+    });
+}
+
+/// Enqueues a newly built task.
+///
+/// Note that the new task is not guranteed to run at once.
+pub(super) fn run_new_task(runnable: Arc<Task>) {
+    // FIXME: remove this check for `SCHEDULER`.
+    // Currently OSTD cannot know whether its user has injected a scheduler.
+    if !SCHEDULER.is_completed() {
+        let fifo_scheduler = Box::new(FifoScheduler::default());
+        let static_scheduler = Box::leak(fifo_scheduler);
+        inject_scheduler(static_scheduler);
+    }
+
+    let should_preempt_info = SCHEDULER
+        .get()
+        .unwrap()
+        .enqueue(runnable, EnqueueFlags::Spawn);
+    if should_preempt_info.is_some() {
+        let cpu_id = should_preempt_info.unwrap();
+        // FIXME: send IPI to set remote CPU's `SHOULD_PREEMPT` if needed.
+        if cpu_id == this_cpu() {
+            SHOULD_PREEMPT.set();
+        }
+    }
+
+    preempt();
+}
+
+/// Dequeues the current task from its runqueue.
+///
+/// This should only be called if the current is to exit.
+pub(super) fn exit_current() {
+    reschedule(&mut |local_rq: &mut (dyn LocalRunQueue + '_)| {
+        if let Some(next) = local_rq.pick_next_current(false) {
+            ReschedAction::SwitchTo(next)
+        } else {
+            ReschedAction::Retry
+        }
+    })
+}
+
+/// Blocks the current task unless `has_woken` is `true`.
+pub(crate) fn park_current(has_woken: &AtomicBool) {
+    reschedule(&mut |local_rq: &mut dyn LocalRunQueue| {
+        if has_woken.load(Ordering::Acquire) {
+            return ReschedAction::DoNothing;
+        }
+
+        local_rq.update_current(UpdateFlags::Wait);
+        if let Some(next) = local_rq.pick_next_current(false) {
```

Comment:
> It should be ok to put the current task to sleep regardless of whether there is the next runnable task.

If there's no next runnable task in the current CPU core, the current control flow will hang regardless of whether `has_woken` is updated if `TryAgain`. Thus, it is necessary to go back and check the flag and perform accordingly to ensure responsiveness and efficiency.

Reply:
I introduced `keep_current` to defer the dequeue action until we find a next, which ensures the `current` is valid and accessible even if we temporarily cannot find a candidate for next `current` (e.g., on returning `ReschedAction::Retry`). This way we keep the invariant that `current` will always be valid once it becomes valid.

If we dequeue first instead and cannot find a next, then the current is supposed to be `None` when we go back to the loop in `reschedule`. Anyone trying to access the `current` of the runqueue will fail there, including a timer interrupt trying to update timing statistics.
```rust
fn reschedule<F>(f: &mut F)
where
    F: FnMut(&mut dyn LocalRunQueue) -> ReschedAction,
{
    let next_task = loop {
        let mut action = ReschedAction::DoNothing;
        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |rq| {
            action = f(rq);
        });

        // act on `action`
    };

    // do something
}
```

Besides, there is another subtle use case in early wake-up scenario. Let's say a sleeping task dequeues the `current` in `park_current` and the first invocation of `pick_next_current` doesn't find a candidate. Then it releases the lock on runqueue and at this time a `wake_up` happens and takes the lock. We know the next invocation of `park_current` would return `ReschedAction::DoNothing` because of the `has_woken` check and the sleeping task will directly become runnable and go on running as `current`. So in the `unpark_current` of this `wake_up`, we need to set the `current` to this task instead of just enqueuing it, which is not supported by our interface.

```rust
// scheduler/mod.rs
pub(crate) fn park_current(has_woken: &AtomicBool) {
    reschedule(&mut |local_rq: &mut dyn LocalRunQueue| {
        if has_woken.load(Ordering::Acquire) {
            return ReschedAction::DoNothing;
        }

        local_rq.dequeue_current();
        local_rq.update_current(UpdateFlags::Wait);
        if let Some(next) = local_rq.pick_next_current() {
            ReschedAction::SwitchTo(next)
        } else {
            ReschedAction::DoNothing // or `ReschedAction::Retry`, which doesn't matter here.
        }
    });
}

// scheduler/mod.rs
pub(crate) fn unpark_target(runnable: Arc<Task>) {
    let should_preempt_info = SCHEDULER
        .get()
        .unwrap()
        .enqueue(runnable, EnqueueFlags::Wake);
    if should_preempt_info.is_some() {
        let cpu_id = should_preempt_info.unwrap();
        // FIXME: send IPI to set remote CPU's `SHOULD_PREEMPT` if needed.
        if cpu_id == this_cpu() {
            SHOULD_PREEMPT.set();
        }
    }
}

// wait.rs
impl Waker {
    pub fn wake_up(&self) -> bool {
        if self.has_woken.swap(true, Ordering::Release) {
            return false;
        }
        scheduler::unpark_target(self.task.clone());

        true
    }

    fn do_wait(&self) {
        let has_woken = &self.has_woken;
        while !has_woken.swap(false, Ordering::Acquire) {
            scheduler::park_current(has_woken);
        }
    }
}
```

Reply:
I introduced `keep_current` flag to defer the actual dequeue action until we find a next. This way we keep an invariant that the `current` will always be valid (non-`None`) once the `current` becomes valid.

If we dequeue first instead and cannot find a next, we have to go back to the loop in `reschedule` with the `current` being `None`. Anyone trying to access `current` will fail there.

```rust
fn reschedule<F>(f: &mut F)
where
    F: FnMut(&mut dyn LocalRunQueue) -> ReschedAction,
{
    let next_task = loop {
        let mut action = ReschedAction::DoNothing;
        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |rq| {
            action = f(rq);
        });

        // act on `action`
    };

    // do something
}
```

Besides, there is another subtle use case for this flag when we `wake_up` a sleeping task after a failed `pick_next_current` in `park_current`. At this time the `current` should be `None`. We know that the sleeping task will directly become runnable and go on running on the next `has_woken` check. Therefore we must not return from `wake_up` with the `current` being `None` but we don't have a mechanism to `set_current` there. A deferred dequeue can solve this problem.

Reply:
> the invariant that current will always be valid once it becomes valid.

Yes. It would be good to have the invariant that a current task can be retrieved at any timing, including in the interrupt context.

> Anyone trying to access the current of the runqueue will fail there, including a timer interrupt trying to update timing statistics.

But I don't think not dequeuing a current, sleeping task is the way to do it. And you have made an excellent case against the idea. If the current task is sleeping but not dequeued (because there is no other runnable tasks), then what will the scheduler tick function do upon the timer interrupts? It will recording the time under this sleeping task's name, which is obviously wrong.

It is possible for the scheduler tick function to fix this time accounting problem by treating this situation specially. But this would only scratch the surface of the more fundamental problem: it is illogical to keep a sleeping task as current and not dequeued it from the runqueue.

So what is the right approach to achieve the invariant that a current task can be retrieved at any timing, including in the interrupt context? I think the answer is idle tasks. If each CPU has an idle task, then there will always be a runnable task.

By the way, even if we don't have such an invariant, the `update_current` function can simply do nothing when there is no current task at the moment. So I think this invariant is a good-to-have, not a must-have.

> there is another subtle use case in early wake-up scenario

I am not sure if I understand your words completely, but the original (draft) code is indeed incorrect. How about the following revised form (it is still just a draft).

```rust
pub(crate) fn park_current(has_woken: &AtomicBool) {
    let mut is_first_try = true;
    reschedule(&mut |local_rq: &mut dyn LocalRunQueue| {
        if is_first_try {
            if has_woken.load(Ordering::Acquire) {
                return ReschedAction::DoNothing;
            }

            local_rq.dequeue_current();
            local_rq.update_current(UpdateFlags::Wait);
        }

        if let Some(next) = local_rq.pick_next_current() {
            ReschedAction::SwitchTo(next)
        } else {
            is_first_try = true;
            ReschedAction::DoNothing
        }
    });
}
```

Reply:
> Thus, it is necessary to go back and check the flag and perform accordingly to ensure responsiveness and efficiency.

The `park_current` method should be designed to minimize the latency of switching to _any_ runnable task, not the _current_ sleeping task again when it becomes runnable. So going all the way back to the loop in `Waiter:: do_wait` does not make sense.

Furthermore, if the CPU is spinning in the loop of `Waiter:: do_wait`, then the scheduler tick function would count all this spinning wait time under the name of this current task. But it shouldn't!

Reply:
@tatetian Thank you for clarifying! The latest commit now dequeue first and retry when we cannot find a next.